File: assiment-1/2.cpp
#include <bits/stdc++.h>
using namespace std;

map<string, vector<string>> graph = {
    {"Raj", {"Priya", "Sunil"}},
    {"Priya", {"Raj", "Aarav", "Neha"}},
    {"Aarav", {"Priya", "Neha", "Arjun"}},
    {"Sunil", {"Raj", "Akash", "Sneha"}},
    {"Akash", {"Sunil", "Neha"}},
    {"Neha", {"Priya", "Akash", "Rahul"}},
    {"Sneha", {"Sunil", "Rahul", "Maya"}},
    {"Rahul", {"Neha", "Sneha", "Pooja"}},
    {"Maya", {"Sneha"}},
    {"Arjun", {"Aarav", "Pooja"}},
    {"Pooja", {"Rahul", "Arjun"}}
};

/* -------- BFS -------- */
void BFS(string start)
{
    set<string> visited;
    queue<string> q;

    q.push(start);
    visited.insert(start);

    cout << "BFS Tree: ";

    while (!q.empty())
    {
        string node = q.front();
        q.pop();
        cout << node << " ";

        for (auto &neighbor : graph[node])
        {
            if (!visited.count(neighbor))
            {
                visited.insert(neighbor);
                q.push(neighbor);
            }
        }
    }
    cout << endl;
}

/* -------- DFS -------- */
void DFS(string node, set<string> &visited)
{
    visited.insert(node);
    cout << node << " ";

    for (auto &neighbor : graph[node])
    {
        if (!visited.count(neighbor))
            DFS(neighbor, visited);
    }
}

int main()
{
    cout << "Social Media Network Traversal\n";

    BFS("Raj");

    cout << "DFS Tree: ";
    set<string> visited;
    DFS("Raj", visited);

    return 0;
}

File: assiment-1/2.py
graph = {
    "Raj": ["Priya", "Sunil"],
    "Priya": ["Raj", "Aarav", "Neha"],
    "Aarav": ["Priya", "Neha", "Arjun"],
    "Sunil": ["Raj", "Akash", "Sneha"],
    "Akash": ["Sunil", "Neha"],
    "Neha": ["Priya", "Akash", "Rahul"],
    "Sneha": ["Sunil", "Rahul", "Maya"],
    "Rahul": ["Neha", "Sneha", "Pooja"],
    "Maya": ["Sneha"],
    "Arjun": ["Aarav", "Pooja"],
    "Pooja": ["Rahul", "Arjun"]
}

start="Raj"
queue=[(start,0)]
level=0;
visited=set(start)
print("breadth-first search:\n")
while queue:
    print(f"level {len(queue)} nodes: {queue}")
    next_level=[]
    for parent in queue:
        if parent[0] not in graph:
            continue
        children=graph[parent[0]]
        for child in children:
            if  child not in visited:
                next_level.append((child,level+1))
                visited.add(child)
    queue=next_level
    level+=1

print("\nDepth-first search:\n")
stack=[start]
reached=set(start)
while(stack):
    print(f"exploring  node {stack[len(stack)-1]}")
    parent=stack.pop()
    if parent not in graph:
        continue
    children=graph[parent]
    for child in children:
        if child not in reached:
            stack.append(child)
            reached.add(child)